flutter 1.22.4 -> 2.0.3, dart 2.10.5 -> 2.12.2

kolaente opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change

Flutter major upgrade to version 2.0.3.

I've upgraded all patches and made sure they apply and build. I did not check if new patches are required due to new version.

I also had to upgrade dart since we need dart in order to build flutter itself. Since the dart sdk is now included in flutter itself, I'm not quite sure if we need this at all, or if it would require changes to the flutter build declaration. To discuss? Maybe someone more knowledgable than me has a better idea :slightly_smiling_face: 

I've been able to create [a new flutter app](https://flutter.dev/docs/get-started/test-drive) with the new built nix flutter package and build it for android.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

kolaente committed:
flutter 1.22.4 -> 2.0.3

r-rmcgibbo commented:
Result of `nixpkgs-review pr 117116` at 3be90191 run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package failed to build (new failures):</summary>
  <ul>
    <li><a href="https://gist.github.com/55408d9d9bda4a07bc6d9a10870a7b82">dart</a></li></li>
  </ul>
</details>
<details>
  <summary>1 suggestion:</summary>
  <ul>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-phase-hooks.md">missing-phase-hooks</a>

`installPhase` should probably contain `runHook preInstall` and `runHook postInstall`.

Near pkgs/development/interpreters/dart/default.nix:49:3:
```
   |
49 |   installPhase = ''
   |   ^
```
</li>
  </ul>
</details>

SuperSandro2000 commented on pkgs/development/interpreters/dart/default.nix:
```diff
@@ -11,24 +11,24 @@
     "${version}-x86_64-darwin" = fetchurl {
       url = "${base}/stable/release/${version}/sdk/dartsdk-macos-${x86_64}-release.zip";
-      sha256 = "1vb2m25w6v901id9syan9q69fa60sxxd7qpyzq21fn5dpah0g99i";
+      sha256 = "0000000000000000000000000000000000000000000000000000";
```
Something went wrong here with hashes.

kolaente replied:
Ah yes. I've noticed it before pushing but forgot to mention after opening the pr.
I just changed the hashes to `000...` for all, then ran `nix-build -A dart` and put the hash from the error message back in as I usually do when updating packages. But since I only have an x86 linux, it only built for that. I guess I need to manually download the files for each arch, check the shasum and put that in the variables?

thiagokokada replied:
Yeah, you need to download each of the files using `nix-prefetch-url <url>`.

It would be nice to have an auto-update script to do this.

kolaente replied:
Should be fixed now.

thiagokokada commented:
> Since the dart sdk is now included in flutter itself, I'm not quite sure if we need this at all

It is included for quite sometime, but we still prefer to build it separately since this way we can put them separately on `environment.systemPackages` (so if someone only needs Dart they don't need to install Flutter too, for example). This also reduces the total derivation size if you need both installed at the same time.

kolaente committed:
dart: fix shasums for darwin x64, linux i686 and linux aarch64

kolaente commented:
@thiagokokada I see, makes sense then. Should I split the flutter and dart update in two separate commits?

Iazel commented:
Hi, I also tried to update this derivation and found a couple of things that would be good to merge in here too:
```diff
diff --git a/mkFlutter.nix b/mkFlutter.nix
index 4a7aa0d..61d1327 100644
--- a/mkFlutter.nix
+++ b/mkFlutter.nix
@@ -30,6 +30,7 @@
 , libXi
 , libXext
 , libGL
+, chromium
 , nspr
 , nss
 , systemd
@@ -45,7 +46,6 @@ let
 
     postPatch = ''
       patchShebangs --build ./bin/
-      find ./bin/ -executable -type f -exec patchelf --set-interpreter $(cat $NIX_CC/nix-support/dynamic-linker) {} \;
     '';
 
     buildPhase = ''
@@ -104,6 +104,9 @@ let
         which
         xz
 
+        # needs chrome for web development
+        chromium
+
         # flutter test requires this lib
         libGLU
 
@@ -137,6 +140,7 @@ runCommand drvName
     #!${bash}/bin/bash
     export PUB_CACHE=''${PUB_CACHE:-"$HOME/.pub-cache"}
     export ANDROID_EMULATOR_USE_SYSTEM_LIBS=1
+    export CHROME_EXECUTABLE=${chromium}/bin/chromium
     ${fhsEnv}/bin/${drvName}-fhs-env ${flutter}/bin/flutter --no-version-check "$@"
   '';
   preferLocalBuild = true;
@@ -155,7 +159,6 @@ runCommand drvName
   };
 } ''
   mkdir -p $out/bin
-
   echo -n "$startScript" > $out/bin/${pname}
   chmod +x $out/bin/${pname}
 ''
```

More in details:
```diff
-      find ./bin/ -executable -type f -exec patchelf --set-interpreter $(cat $NIX_CC/nix-support/dynamic-linker) {} \;
```
All executables files in `./bin` are actually shell scripts and `patchelf` will complain about it during compilation.

```diff
+    export CHROME_EXECUTABLE=${chromium}/bin/chromium
```
Flutter 2 will now have web development as a compilation target, hence `flutter doctor` will complain about it if no browser is present. `chromium` here looks like a good choice to avoid enable `allowUnfree`.

Hope it was useful! ;)

tyrion commented:
Anyway, there is an issue with the current approach of separating dart and flutter. That is, the dart compiler cannot find flutter by default (i.e. if you add both flutter and dart to a mkShell).

This is sometimes needed, for example when running

```
$ dart pub outdated --mode=null-safety
Package doesn't exist (the Flutter SDK is not available).
```

Running the same command as `flutter pub outdated --mode=null-safety` works.

Is there an easy way to "enter" the flutter's fhsEnv? That would allow to have a working dart and could also be handy for development, I believe.

thiagokokada commented:
> Hi, I also tried to update this derivation and found a couple of things that would be good to merge in here too:
> 
> ```diff
> diff --git a/mkFlutter.nix b/mkFlutter.nix
> index 4a7aa0d..61d1327 100644
> --- a/mkFlutter.nix
> +++ b/mkFlutter.nix
> @@ -30,6 +30,7 @@
>  , libXi
>  , libXext
>  , libGL
> +, chromium
>  , nspr
>  , nss
>  , systemd
> @@ -45,7 +46,6 @@ let
>  
>      postPatch = ''
>        patchShebangs --build ./bin/
> -      find ./bin/ -executable -type f -exec patchelf --set-interpreter $(cat $NIX_CC/nix-support/dynamic-linker) {} \;
>      '';
>  
>      buildPhase = ''
> @@ -104,6 +104,9 @@ let
>          which
>          xz
>  
> +        # needs chrome for web development
> +        chromium
> +
>          # flutter test requires this lib
>          libGLU
>  
> @@ -137,6 +140,7 @@ runCommand drvName
>      #!${bash}/bin/bash
>      export PUB_CACHE=''${PUB_CACHE:-"$HOME/.pub-cache"}
>      export ANDROID_EMULATOR_USE_SYSTEM_LIBS=1
> +    export CHROME_EXECUTABLE=${chromium}/bin/chromium
>      ${fhsEnv}/bin/${drvName}-fhs-env ${flutter}/bin/flutter --no-version-check "$@"
>    '';
>    preferLocalBuild = true;
> @@ -155,7 +159,6 @@ runCommand drvName
>    };
>  } ''
>    mkdir -p $out/bin
> -
>    echo -n "$startScript" > $out/bin/${pname}
>    chmod +x $out/bin/${pname}
>  ''
> ```
> 
> More in details:
> 
> ```diff
> -      find ./bin/ -executable -type f -exec patchelf --set-interpreter $(cat $NIX_CC/nix-support/dynamic-linker) {} \;
> ```
> 
> All executables files in `./bin` are actually shell scripts and `patchelf` will complain about it during compilation.
> 
> ```diff
> +    export CHROME_EXECUTABLE=${chromium}/bin/chromium
> ```
> 
> Flutter 2 will now have web development as a compilation target, hence `flutter doctor` will complain about it if no browser is present. `chromium` here looks like a good choice to avoid enable `allowUnfree`.
> 
> I also noticed that in `mkDerivation`, we do not specify the `version`, which should make the upgrade system not work as expected. I will therefore suggest to also add an:
> 
> ```
> inherit patches version;
> ```
> 
> Hope it was useful! ;)

I think that adding `chromium` to the derivation is overkill. If you want to use Chrome you can export `CHROME_EXECUTABLE` environment variable yourself, and this is also good because maybe you don't have Chromium but have Google Chrome/Brave/Any other Chromium based browser installed, and can use what you want.

But yeah, maybe it would be better to document this somewhere.

> Anyway, there is an issue with the current approach of separating dart and flutter. That is, the dart compiler cannot find flutter by default (i.e. if you add both flutter and dart to a mkShell).
> 
> This is sometimes needed, for example when running
> 
> ```
> $ dart pub outdated --mode=null-safety
> Package doesn't exist (the Flutter SDK is not available).
> ```
> 
> Running the same command as `flutter pub outdated --mode=null-safety` works.
> 
> Is there an easy way to "enter" the flutter's fhsEnv? That would allow to have a working dart and could also be handy for development, I believe.

I think using `flutter pub` instead is a good workaround no :thinking: ?

kolaente committed:
flutter: add inherit version

kolaente commented:
I've added the `inherit version` and removed the find command, looks like it works fine (building an app worked without issues, but I obviously didn't test everything).

> But yeah, maybe it would be better to document this somewhere.

Actually, if you run `flutter doctor` without chrome specified it will tell you to set `CHROME_EXECUTABLE`:
```
Doctor summary (to see all details, run flutter doctor -v):
[✓] Flutter (Channel stable, 2.0.3, on Linux, locale en_GB.UTF-8)
[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.3)
[✗] Chrome - develop for the web (Cannot find Chrome executable at google-chrome)
    ! Cannot find Chrome. Try setting CHROME_EXECUTABLE to a Chrome executable.
[✓] Android Studio (version 4.0)
[!] Connected device
    ! No devices available

! Doctor found issues in 2 categories.
```

So I think it'll be fine if we don't really document this as flutter already made that env variable easily discoverable.

kolaente committed:
flutter: remove patchelf

Iazel commented:
Ah, not including `chromium` is indeed better! I thought that FHS meant it will be some sort of isolated environment ala `nix-shell --pure`, but I definitely misunderstood it xD

roosemberth commented:
This is important to me, what needs to be done so that this MR can be merged?

kolaente commented:
@roosemberth A review from at least @babariviere would be great 🙂 

If you need to, you could also just clone my branch and build flutter yourself and use that for now.

SuperSandro2000 commented:
This is a semi-automatic executed nixpkgs-review with [nixpkgs-review-checks extension](https://github.com/SuperSandro2000/nixpkgs-review-checks). It is checked by a human on a best effort basis and does not build all packages (e.g. lumo, tensorflow or pytorch).
If you have any questions or problems please reach out to SuperSandro2000 on IRC.

Result of `nixpkgs-review pr 117116` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
<summary>
3 packages built:
</summary>
<ul>
<li>dart</li>
<li>flutter (flutterPackages.stable)</li>
<li>hover</li>
</ul>
</details>

SuperSandro2000 reviewed (approved):
package wise this is fine

ericdallo reviewed (approved):
LGTM

thiagokokada commented:
Package wise it seems good, still needs to test.

thiagokokada commented:
Tested:

```
$ flutter creater new_project
Your application code is in new_project/lib/main.dart.

$ flutter build web
Compiling lib/main.dart for the Web...                             18.6s

$ flutter run
Unable to run "adb", check your Android SDK installation and ANDROID_SDK_ROOT environment variable:
/nix/store/divwb1bx12x84nps3iak4izd0w37569v-platform-tools-29.0.6/libexec/android-sdk/platform-tools/adb

$ flutter build apk
/nix/store/0w6dqz95lgdwgnp24cc95m184g7yhgka-openjdk-11.0.9+11/lib/openjdk/bin/java: symbol lookup error: /usr/lib/libc.so.6: undefined symbol: _dl_fatal_printf, version GLIBC_PRIVATE
Running Gradle task 'assembleRelease'...
Running Gradle task 'assembleRelease'... Done                       71ms
Gradle task assembleRelease failed with exit code 127
```

Hover also seems broken, but maybe we could merge this and fix it later?

thiagokokada reviewed (approved):
Some things are broken, but maybe it is better to merge this PR as-is and work to fix what is broken later.

ericdallo commented:
Which things are broken?

thiagokokada commented:
> Which things besides hover are broken?

Just see my tests, `flutter build apk` is broken, `flutter run` also seems to not be working (this last one maybe is the lack of an connected Android device in my system, since adb itself works fine).

thiagokokada commented:
BTW, `flutter build apk` is maybe broken because I am running `release-20.09`, and I am not sure that this package is hermetic so maybe it is leaking somewhere.

ericdallo commented:
I see, so I'm fine in merging and fix the rest later

thiagokokada commented:
About the `hover` issue: I think it was already broken before, it is just that I never needed to run `hover init`.

thiagokokada commented:
> About the `hover` issue: I think it was already broken in `master`, it is just that I never needed to run `hover init`.

Yeah, the issue with `hover` is completely unreleated: https://github.com/NixOS/nixpkgs/pull/118187

kolaente commented:
@thiagokokada interesting `flutter build apk` did not work for you as I was able to build an apk just fine. What project were you trying to build?

roosemberth commented:
I am able to run `flutter run` and `flutter build apk` on the provided branch (cf. attached flake).

PS: Github won't allow me to attach a file, so I've pasted it on the details section:

<details>
```
{
  inputs.nixpkgs.url = "github:kolaente/nixpkgs?ref=ba619289b62453a4957e540c109880d5d261ce36";

  outputs = { self, nixpkgs }: let
    systems = [ "x86_64-linux" "i686-linux" ];
    forAllSystems = f: nixpkgs.lib.genAttrs systems (system: f (import nixpkgs {
      inherit system; overlays = [ self.overlay ];
      config.android_sdk.accept_license = true;
    }));
  in {
    overlay = final: prev: with final; {
      android = androidenv.composeAndroidPackages {
        platformVersions = [ "28" ];
        abiVersions = [ "x86_64" ];
        buildToolsVersions = [ "29.0.2" ];
      };
      emulator = androidenv.emulateApp {
        name = "emulator";
        enableGPU = true;
        platformVersion = "28";
        abiVersion = "x86_64";
      };
    };

    devShell = forAllSystems (pkgs: with pkgs; stdenv.mkDerivation {
      name = "android-env";
      ANDROID_HOME = "${android.androidsdk}/libexec/android-sdk";
      ANDROID_JAVA_HOME = "${jdk.home}";
      JAVA_HOME = "${jdk.home}";
      nativeBuildInputs = [ android.androidsdk emulator flutter ];
      allowSubstitutes = false;
    });

    # Export all packages from the overlay attribute.
    packages = forAllSystems (pkgs: with pkgs.lib;
      getAttrs (attrNames (self.overlay {} {})) pkgs
    );

    apps = forAllSystems (pkgs: with pkgs; {
      emulator.type = "app";
      emulator.program = "${emulator}/bin/run-test-emulator";
    });
  };
}
```
</details>

roosemberth commented:
Hooray! :tada:

thiagokokada commented:
@kolaente Like I said, I think this package is not hermetic enough. This GNU GLIBC error probably means that there is a mismatch between GLIBC versions (I am using `release-20.09`, I think most people are testing this in unstable).

This is something that should be eventually fixed, since non hermetic packages are bad since they're non reproducible.

roosemberth commented:
@thiagokokada How about working that out in a new issue describing what is expected and how it differs from the result under what circunstances?
